numpy_financial/tests/test_financial.py

```diff
@@ -329,6 +215,129 @@ def test_no_interest(self):
         assert_(npf.nper(0, -100, 1000) == 10)
 
 
+class TestPpmt:
+    def test_float(self):
+        assert_allclose(
+            npf.ppmt(0.1 / 12, 1, 60, 55000),
+            -710.25,
+            rtol=1e-4
+        )
+
+    def test_decimal(self):
+        result = npf.ppmt(
+            Decimal('0.1') / Decimal('12'),
+            Decimal('1'),
+            Decimal('60'),
+            Decimal('55000')
+        )
+        assert_equal(
+            result,
+            Decimal('-710.2541257864217612489830917'),
+        )
+
+    @pytest.mark.parametrize('when', [1, 'begin'])
+    def test_when_is_begin(self, when):
+        assert_allclose(
+            npf.ppmt(0.1 / 12, 1, 60, 55000, 0, when),
+            -1158.929712,  # Computed using Google Sheet's PPMT
+            rtol=1e-9,
+        )
+
+    @pytest.mark.parametrize('when', [None, 0, 'end'])
+    def test_when_is_end(self, when):
+        args = (0.1 / 12, 1, 60, 55000, 0)
+        result = npf.ppmt(*args) if when is None else npf.ppmt(*args, when)
+        assert_allclose(
+            result,
+            -710.254126,  # Computed using Google Sheet's PPMT
+            rtol=1e-9,
+        )
+
+    @pytest.mark.parametrize('when', [Decimal('1'), 'begin'])
+    def test_when_is_begin_decimal(self, when):
+        result = npf.ppmt(
+            Decimal('0.08') / Decimal('12'),
+            Decimal('1'),
+            Decimal('60'),
+            Decimal('15000.'),
+            Decimal('0'),
+            when
+        )
+        assert_almost_equal(
+            result,
+            Decimal('-302.131703'),  # Computed using Google Sheet's PPMT
+            decimal=5,
+        )
+
+    @pytest.mark.parametrize('when', [None, Decimal('0'), 'end'])
+    def test_when_is_end_decimal(self, when):
+        args = (
+            Decimal('0.08') / Decimal('12'),
+            Decimal('1'),
+            Decimal('60'),
+            Decimal('15000.'),
+            Decimal('0')
+        )
+        result = npf.ppmt(*args) if when is None else npf.ppmt(*args, when)
+        assert_almost_equal(
+            result,
+            Decimal('-204.145914'),  # Computed using Google Sheet's PPMT
+            decimal=5,
+        )
+
+    @pytest.mark.parametrize('args', [
+        (0.1 / 12, 0, 60, 15000),
+        (Decimal('0.012'), Decimal('0'), Decimal('60'), Decimal('15000'))
+    ])
+    def test_invalid_per(self, args):
+        # Note that math.isnan() handles Decimal NaN correctly.
+        assert math.isnan(npf.ppmt(*args))
+
+    @pytest.mark.parametrize('when, desired', [
+        (
+            None,
+            [-75.62318601, -76.25337923, -76.88882405, -77.52956425],
+        ), (
+            [0, 1, 'end', 'begin'],
+            [-75.62318601, -75.62318601, -76.88882405, -76.88882405],
+        )
+    ])
+    def test_broadcast(self, when, desired):
```

Comment:
This looks like it's testing only the first slice of the expected array (i.e. period 1) not each period for each when. I realise that the desired values are from the original tests, and that this is a refactoring, but I would prefer not to propagate an incorrect test.

i.e. I would expect the following two results to be the same, but they are not.

```python
In [7]: whens = [0, 1, 'begin', 'end']                                                                                                                                                                                  

In [8]: npf.ppmt(*args, whens)                                                                                                                                                                                          
Out[8]: array([-75.62318601, -75.62318601, -76.25337923, -77.52956425])

In [9]: for when in whens: 
   ...:     print(npf.ppmt(*args, when))                                                                                                                                                                                                            
[-75.62318601 -76.25337923 -76.88882405 -77.52956425]
[-91.52712662 -75.62318601 -76.25337923 -76.88882405]
[-91.52712662 -75.62318601 -76.25337923 -76.88882405]
[-75.62318601 -76.25337923 -76.88882405 -77.52956425]
```

Am I missing something obvious here, or are the existing tests incorrect?

Reply:
Oh, IIUC you would only get the second result if you did something like `when = [[0], [1], ['end'], ['begin']]` or `numpy.arange(1, 5).reshape(4, 1)`. For example:

```python
>>> import numpy
>>> import numpy_financial as npf
>>> whens = [0, 1, 'end', 'begin']
>>> npf.ppmt(0.1 / 12, numpy.arange(1, 5).reshape(4, 1), 24, 2000, 0, whens)
array([[-75.62318601, -91.52712662, -75.62318601, -91.52712662],
       [-76.25337923, -75.62318601, -76.25337923, -75.62318601],
       [-76.88882405, -76.25337923, -76.88882405, -76.25337923],
       [-77.52956425, -76.88882405, -77.52956425, -76.88882405]])
```

That *is* a much better broadcasting test than the one that exists now.

Reply:
That is what I had in mind, in many cases that would also much more useful. This also raises the question of what happens if more variables are included e.g. we should expect a 2 x 4 x4 array if we use rate = [0.1 / 12, 0.2 / 12].

Reply:
As this PR is for cleaning up - not changing the existing behaviour - I'm happy to mark this as resolved.